cookbook/x-flat-jwt.md

```diff
@@ -0,0 +1,324 @@
+      stripEndpoint: true
+```
+
+Sending a request to FLAT running on localhost port 1234 results in:
+
+```
+curl -i http://localhost:1234/httpbin/anything
+HTTP/1.1 200 OK
+…
+Content-Type: application/json
+
+{
+  "args": {},
+  "data": "",
+  "files": {},
+  "form": {},
+  "headers": {
+    "Accept": "*/*",
+    "Accept-Encoding": "deflate, gzip",
+    "Host": "httpbin.org",
+    "User-Agent": "curl/7.29.0",
+    "X-Amzn-Trace-Id": "…" 
+  },
+  "json": null,
+  "method": "GET",
+  "origin": "…",
+  "url": "https://httpbin.org/anything" 
+}
+```
+
+That's what you would expect from httpbin.org, right?
+
+## Restricting access: Swagger Security and `x-flat-jwt`
+
+Now, you don't want anyone except authorized users to use this proxy. This is typically achieved with access tokens. Some tokens are JSON Web Tokens ([JWT](https://tools.ietf.org/html/rfc7519)), while others are opaque.
+
+Swagger has a two-part feature to describe protected access to routes: `securityDefinitions` (what sort of protection is applied …) and `security` (… to which routes), e.g.:
+
+```yaml
+swagger: "2.0" 
+basePath: /
+securityDefinitions:
+  JWTCookie:
+    type: apiKey
+    in: header
+    name: Cookie
+paths:
+  /httpbin/**:
+    security:
+      - JWTCookie: []
+```
+
+This defines a security scheme object (named `JWTCookie`), meaning that some sort of cookie is needed to access certain routes. This is applied to the [wildcard path](/reference/OpenAPI/differences.md#wildcard-paths) `/httpbin/**`.
+
+This documentation feature, with some extensions, is used to make FLAT actually permit access to the route only if a valid JWT token is presented.
+
+First, we define the name of the cookie expected to accompany the API request:
+
+```yaml
+…
+securityDefinitions:
+  JWTCookie:
+    type: apiKey
+    in: header
+    name: Cookie
+    x-flat-cookiename: authtoken    # ⬅ specify the cookie name
+…
+```
+
+Then we specify the [configuration](/reference/OpenAPI/security.md#the-x-flat-jwt-field) for decoding the JWT token:
+
+```yaml
+…
+  JWTCookie:
+    type: apiKey
+    in: header
+    name: Cookie
+    x-flat-cookiename: authtoken
+    x-flat-jwt:                     # ⬅ our JWT configuration:
+      key:                          # ⬅ the key to decode the JWT …
+        file: secret.pem            # ⬅ … is read from the file secret.pem
+      alg: RS256                    # ⬅ the signing algorithm is RS256
+…
+```
+
+The specified key is a public key, read from the file secret.pem, e.g.:
+```
+-----BEGIN PUBLIC KEY-----
+MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDGSd+sSTss2uOuVJKpumpFAaml
+t1CWLMTAZNAabF71Ur0P6u833RhAIjXDSA/QeVitzvqvCZpNtbOJVegaREqLMJqv
+FOUkFdLNRP3f9XjYFFvubo09tcjX6oGEREKDqLG2MfZ2Z8LVzuJc6SwZMgVFk/63
+rdAOci3W9u3zOSGj4QIDAQAB
+-----END PUBLIC KEY-----
+```
+
+That's all. Now FLAT will only permit requests if they supply a token that bear an RS256 signature that was created with the private key that matches the given public key.
+
```

Comment:
```suggestion

Usually, you would get the key and algorithm from your identity provider (e.g. an OAuth2 server). That service would be responsible for issuing JWT tokens for your users.

For this tutorial we have prepared a couple of JWT tokens for you to try out different situations.

Valid:

eybGciOiJSUzI1NiJ9.eyJzdWIiOiJzb21lX3VzZXIiLCJpc3MiOiJzb21lX3Byb3ZpZGVyIn0.bNXv28XmnFBjirPbCzBqyfpqHKo6PpoFORHsQ-80IJLi3IhBh1y0pFR0wm-2hiz_F7PkGQLTsnFiSXxCt1DZvMstbQeklZIh7O3tQGJyCAi-HRVASHKKYqZ_-eqQQhNr8Ex00qqJWD9BsWVJr7Q526Gua7ghcttmVgTYrfSNDzU

Expired:
eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJzb21lX3VzZXIiLCJpc3MiOiJzb21lX3Byb3ZpZGVyIiwiZXhwIjoxNTkwNDkxNTI4fQ.lJnUpBzMx84_5yigeHeLw4f8sbdSdu_7fWr1--t7EAp8v8K-kSmVYUGnR0Jx4o_ZE84N2M72Kn1pKssrzgTHsFi7txcZHHz_JqgnPgKqsZwjrmWDC-XVvdrSXjAsPO6wn0qy3KEMT1y6Z8YQA4ZyzA1dDsRRIUFiNrgF6_b5pC4


```

Reply:
i couldn't add markdown code markers in the "suggest" box :(.

with this section, we could make it clearer where the tokens come from that we use in the following examples.

or do you think this will make it too long?

Reply:
If you tell the reader up-front that one of the tokens is expired, there's no surprise when using it. :-)